### ISSUE TITLE
Bug: order-by=file is compiler-dependent

### ISSUE THREAD

**Issue by DaanDeMeyer** [onqtam/doctest#287]
See https://github.com/DaanDeMeyer/doctest/runs/216609947.

Even more weird is that the output not only differs from the rest of the compilers but the output also differs between cl and clang-cl as well.

**Comment by onqtam** [comment]
I'll take a look as soon as possible but it could take 1-2 days... sorry.

**Comment by onqtam** [comment]
I noticed now that its not only the clang-cl builds which fail but also some cl builds: [```windows-2019-cl```](https://github.com/DaanDeMeyer/doctest/runs/216609972). If this was only a clang-cl problem I would have pointed you to check if [this](https://github.com/onqtam/doctest/blob/2.3.4/doctest/parts/doctest.cpp#L1048) is the reason but if it happens with cl as well I'm clueless for now...

**Comment by DaanDeMeyer** [comment]
When using `clang-cl`, `__FILE__` expands to `..\\examples\\all_features/header.h` for `header.h` and `..\\examples\\all_features\\main.cpp` for all .cpp files.

When using `clang++` (on Windows), `__FILE__` expands to `../examples/all_features/header.h` for `header.h` and `../examples/all_features/main.cpp` for all .cpp files.

When using `cl`, `__FILE__` expands to `C:\\Users\\daan\\projects\\doctest\\examples\\all_features\\header.h` for `header.h` and `..\\examples\\all_features\\main.cpp` for all .cpp files.

So basically, we're going to have to do some sort of path normalization if we want reliable results from `FileOrderComparator`.

**Comment by onqtam** [comment]
Gosh... such "exciting" software engineering problems!
This issue is related to path normalization (on which I gave up on...): https://github.com/onqtam/doctest/issues/45

one thing we could do is to use either a new hidden flag or ```no_path_in_filenames``` and something like [```skipPathFromFilename()```](https://github.com/onqtam/doctest/blob/2.3.4/doctest/parts/doctest.cpp#L636) so the output gets stable across compilers at least for testing doctest itself, and we could log a separate issue for fixing this properly later.

One of the pros of not normalizing paths (and using something like ```skipPathFromFilename()```) is that we don't need to allocate memory for the filename strings - doctest just uses a ```const char*``` pointing to the memory in the binary with the constant.

**Comment by onqtam** [comment]
one more detail: the reason only ```filter_2``` fails is because it is one of the rare tests which aren't for the [specific .cpp file](https://github.com/onqtam/doctest/blob/master/examples/all_features/CMakeLists.txt#L41) - as it can be seen [here](https://github.com/onqtam/doctest/blob/master/examples/all_features/CMakeLists.txt#L68) only that test can span multiple sources/headers (doesn't specify the ```--source=file=<...>``` option).

**Comment by DaanDeMeyer** [comment]
Is it an option to change the default `order_by` value and just document the problem? 

This feels like an edge case without a clean solution (at least until we can think of something) so I'd rather just avoid it from occurring for now.

**Comment by onqtam** [comment]
indeed I think changing the order to "name" or "suite" will do the trick!

And we should rename this issue to reflect the actual problem.

**Comment by DaanDeMeyer** [comment]
Should I change it just for the test or the default value? I think changing the default value is "the right thing to do" as the current behaviour is broken but users might be relying on the default order so it would be a breaking change.

I think changing the order just for the test is the safest thing for now. I'll also add documentation to the `order-by` option that it's compiler dependent.

**Comment by onqtam** [comment]
yes, changing it for the test only should be the safest thing to do.

**Comment by onqtam** [comment]
We could still keep this open, since the bug is still present..?

**Comment by DaanDeMeyer** [comment]
Of course, my bad.

**Comment by maidenone** [comment]
Ok this is horrible, but it solved our problem. Putting this here for others with simular problems.

We are compiling using cmake + clang + ninja on windows.

in the cmake file we declare include folders with: 

`target_include_directories(target PUBLIC "/proj/include/")`

In our project, when we include a .hpp file from a .cpp file doctest get:

`/proj/include/file.hpp`

but if we include a .hpp from another .hpp it results in:

`/proj/include\file.hpp`

As we put tests in file.hpp, and include from .cpp and .hpp files those tests will run twice.
Our solution to this was to rewrite the path in the test_case constructor:

```c++
        m_type        = type;                         // line 4255 in doctest.h
        m_template_id = template_id;        // line 4256 in doctest.h

        for (size_t i = 0; i < m_file.size(); i++)
        {
            if (m_file[i] == '\\')
            {
                m_file[i] = '/';
            }
        }
```

**Comment by cdeln** [comment]
@maidenone I don't understand how your problem is related to this issue. Please provide a minimal reproducible example to clarify.

**Comment by cdeln** [comment]
Unless we get more input on this I suggest we close since the issue is documented in the reference.
